### PR TITLE
Reinstating scphylo Ancestor-Descendant Metric

### PR DESCRIPTION
**Pull request by gordonkoehn**
After a brief exchange with Farid from scphylo I have decided to reinstate his implementation of the AD distance. 
See our helpful exchange here:
- https://github.com/faridrashidi/scphylo-tools/issues/8

It appears that AD is not uniquely defined. Scphylo considers only pairs of mutations, whilst I considered all node pairs including the root node. The literature does not seem explicit about this. See best explanation here https://dl.acm.org/doi/10.1145/3233547.3233584

The scphylo metric throws errors if the first tree is a star tree, i.e. having no ancestor-descendant pairs. Farid, said he'll catch this behaviour. For now, I added a `try catch` and return a `nan` as in those cases no AD exists.

 Laura considers the AD with the root, yet this effectively adds a bias for us by comparing if nodes exist in both trees. 
 As we have all mutations present in our simulated trees, I reinstated the scphylo metric. 
 
 This means that we won't have distances for the AD if the star tree is the initial tree. - Unfortunate, but rather we have different distances to compensate.

### CODE REVIEW

File: src/pyggdrasil/distances/_scphylo_wrapper.py [hunk 1]
```diff
@@ -32,7 +36,12 @@ def calculate(self, /, tree1: anytree.Node, tree2: anytree.Node) -> float:
 
         df1 = utils.tree_to_dataframe(tree1)
         df2 = utils.tree_to_dataframe(tree2)
-        return scphylo.tl.ad(df1, df2)
+        try:
```
Comment (pawel-czyz) [inline src/pyggdrasil/distances/_scphylo_wrapper.py:39]: I guess raising the error is better:

```
try:
    return scphylo.tl.ad(df1, df2)
except ZeroDivisionError:
    raise ZeroDivisionError("Error known to happen when no pairs of mutations exist in input tree. Consider <the name of the improved metric>") 
```

In particular, NaNs could be propagated further down and then e.g., averages or some plots will have some unexpected behaviour.

**Review by pawel-czyz — APPROVED** [review]
Very nice work and great that you submitted your very first issue at another repository!

I'd suggest to keep raising the error, so that we don't have troubles with NaN's down in the pipeline.

File: src/pyggdrasil/distances/_scphylo_wrapper.py (hunk 1 above)
Comment (pawel-czyz) [reply src/pyggdrasil/distances/_scphylo_wrapper.py:39]: Good point! What do you think about something like this?
 
```
class AncestorDescendantSimilarity(interface.TreeSimilarity):
    def __init__(self, raise_error: bool) -> None:
        self._raise_error = raise_error
    
    ...

        try:
            distance = scphylo.tl.ad(df1, df2)
        except ZeroDivisionError:
            if self._raise_error:
               raise ValueError("Error known to happen when no pairs of mutations exist in input tree.")
            else:
                return jnp.nan
```

Comment (gordonkoehn) [reply src/pyggdrasil/distances/_scphylo_wrapper.py:39]: Yes - that surely works. 

Actually, that is just what Farid just did in this [commit](https://github.com/faridrashidi/scphylo-tools/commit/6f3c65cd41c80926eefa700735f85a71bc42ca0c) 

Perhaps we should update our fork  - Or is it easier to just do this ourselves? 

But that still leaves the question of what the snakemake workflow is supposed to do when calculating distances of star trees. The two options seem to be:

1) excluding them from the workflow - i.e. not generating the files.  - implies a somewhat cumbersome output filepath generation for all star tree / AD experiments
2) generate `nan` and empty plots

I am inclined to the first option.

File: workflows/mark01.smk [hunk 2]
```diff
@@ -56,18 +56,21 @@ def make_all_mark01()->list[str]:
...
+                tree_id = TreeId(tree_type=TreeType(tree_type), n_nodes=n_node, seed=tree_seed)
+
+                for n_cell in n_cells:
+                        for error_name, error in errors.items():
+                            for metric in metrics:
+                                # AD is not defined for star trees - skip this case
+                                if tree_type == "s" and metric == "AD":
+                                    continue
+                                filepaths.append(filepath+f"{tree_id}-{n_cell}_{error['fpr']}_{error['fnr']}_0.0_{observe_homozygous}_{cell_attachment_strategy}/{metric}_hist.svg")
+
```
Comment (gordonkoehn) [inline workflows/mark01.smk:73]: To avoid a failing snakemake workflow I have not skipped all special cases of the initial star tree and AD metric. 

Can do this from now on - just means one think to take care of when using the Ad.

File: src/pyggdrasil/distances/_scphylo_wrapper.py (hunk 1 above)
Comment (pawel-czyz) [reply src/pyggdrasil/distances/_scphylo_wrapper.py:39]: As I haven't touched the fork for quite some time, I worry that it may have diverged by quite a lot and trying to update it would lead to all sorts of compatibility issues :sweat_smile: 

As your solution below works we can drop modifying our A-D distance as well :slightly_smiling_face:

**Review by pawel-czyz — APPROVED** [review]
I think this works for us, thanks for investigating the issue!
